_posts/2016-08-26-babili.md

```diff
@@ -0,0 +1,183 @@
+
+## Example
+
+When it's possible to only target browsers that support newer ES features, code sizes can be smaller because you don't have to transpile and then minify.
+
+For example: say we want to minify a file that contains a ES6 class and we want to target the latest version of the chrome.
+
+```js
+class Mangler {
+  constructor(program) {
+    this.program = program;
+  }
+}
+// need this since otherwise Mangler isn't used
+new Mangler(); 
+```
+
+Before, we might run Babel to transpile [the class into a function](http://babeljs.io/docs/plugins/transform-es2015-classes) and run uglify on the compiled code to send to the browser.
+
+```js
+// ES6 code -> Babel -> Uglify/Babili -> Minified ES5 Code
+var Mangler=function a(b){_classCallCheck(this,a),this.program=b};Mangler();
+```
+
+With the babel minifier, you can just run the minifier which will work on ES6 code.
+
+```js
+// ES6 code -> Babili -> Minified ES6 Code
+class a{constructor(b){this.program=b}}new a;
+```
+
+Also it's important to note that this isn't specific to ES6. Because Babel updates as ECMAScript updates (with [ES2015, ES2016, and now ES2017](http://babeljs.io/docs/plugins/#official-presets)) and follows the proposal process for experimental features (with our [stage-x presets](http://babeljs.io/docs/plugins/#stage-x-experimental-presets)), the minifier should be able to output whatever version of ECMAScript that is supported.
+
+In addition, there are also specific optimizations we can make when we minify ES2015+ code.
+
+## Usage
+
+### Babel Preset
+
+If you already use Babel, you can just add the [babili](https://github.com/babel/babili#babel-preset) preset (babel-preset-babili) to your config.
+
+You probably want to enable this only in production, so use the [env option](http://babeljs.io/docs/usage/babelrc/#env-option) which uses either `process.env.BABEL_ENV` or `process.env.NODE_ENV`
+
+```bash
+$ npm install babel-preset-babili --save-dev
+```
+
+```js
+// previous .babelrc
+{ "presets": ["es2015"] }
+// .babelrc
+{
+  "presets": ["es2015"],
+  "env": {
+    "production": {
+      "presets": ["babili"]
+    }
+  }
+}
+```
+
+### Babili CLI
+
+If you don't use Babel, you can use our standalone cli tool [`babili`](https://github.com/babel/babili#cli). (Currently it's just a wrapper for babel-cli + the preset). You could run this after transpiling (or not) in place of uglify.
+
+```bash
+$ npm install babili --save-dev
+```
+
+```bash
+# equivalent to 
+# babel src -d lib --presets=babili
+$ babili src -d lib
+```
+
+### Webpack
+
+You can either just use the preset option above with `babel-loader`, or use it seperately with the [babili-webpack-plugin](https://github.com/boopathi/babili-webpack-plugin) (made by [@boopathi](https://github.com/boopathi/), who also works on babili).
```

Comment:
The current build sequence is `transpile -> bundle -> minify` ... As we don't traverse through Objects and its properties, I think there will not be much optimizations on the bundler generated code other than mangling and whitespace/comments removal. It can be made `(transpile + minify) -> bundle` if the bundler already produces mangled and has an option to output minified (just the syntax and comments) code. So I was thinking in terms of,

1. transpile + minify
2. bundle - remove dead code outside module level (tree-shaking?)
3. `transform(bundledcode, {minified: true, compact: true, comments: false, plugins: ["mangle"]})` as the last step after bundling.

So, I'm not sure if such a webpack plugin should exist. I used it for testing and just polished and published so others can try babili in the current toolchain quick. If it's worth maintaining, I'm up for moving it to babel org.